### PR TITLE
read and write methods for nn.Module.

### PR DESCRIPTION
**Pull request by dominikgrewe**
Allows modules that want to implement custom `read` and `write` methods to
call `parent.read/write`.

### CODE REVIEW

**Comment by nicholas-leonard** [comment]
@dominikgrewe This won't break anything, but now all modules will be serialized through those methods you just created.

**Comment by dominikgrewe** [comment]
Yes, but they're just doing what torch serialization would have done anyway (apart from filtering out non-writable objects).

**Comment by soumith** [comment]
looks pretty useful! Thanks.

**Comment by soumith** [comment]
@nicholas-leonard this is useful when implementing custom read-write, you can change some fields and then just call parent.read /write

**Comment by eladhoffer** [comment]
@soumith can't we use this to avoid the need for #526 by adding an option to write just weight, bias and running_mean/var (empty tensors for all other fields)?

**Comment by dominikgrewe** [comment]
There are various other things you might want to save though, e.g. the convolution settings (kernel size, stride, padding etc) for a convolution module. So we could only do without `clearState` if many modules had custom read & write functions. It's worth thinking about though. What other uses for `clearState` are there aside from serialization?

**Comment by eladhoffer** [comment]
@dominikgrewe yes, what I meant is removing the output, gradInput and buffer tensors. Other fields should be kept.

**Comment by soumith** [comment]
@eladhoffer not a good idea. if the intention is to save a model, it's to save the whole model as is. we can introduce an nn.save that calls clearState before hand, and that's okay.
